test/function/236.c

```diff
@@ -0,0 +1,144 @@
+ * MPS, posts on a semaphore to synchronize with the main thread, and
+ * then blocks itself in a read() system call. The main thread then
+ * calls mps_arena_collect() to cause the MPS to suspend all other
+ * threads. If the issue is not fixed, then the read() call on the
+ * thread fails with EINTR. Alternatively, if the issue is fixed, the
+ * read() call is automatically restarted, the main thread writes to
+ * the other end of the pipe so that the read() call completes, and we
+ * tear everything down.
+ */
+
+typedef struct test_data_s {
+  mps_arena_t arena;            /* The arena. */
+  sem_t *sem;                   /* Semaphore to post once registered. */
+  int fd;                       /* File descriptor to read from. */
+} test_data_t;
+
+static void *thread_fn(void *arg)
+{
+  test_data_t *data = arg;
+  mps_thr_t thread;
+  int e;
+  char c;
+
+  die(mps_thread_reg(&thread, data->arena), "mps_thread_reg");
+  e = sem_post(data->sem);
+  asserts(e == 0, "sem_post: %s", strerror(errno));
+
+  /* There is a race here: if MPS goes ahead with the collection on
+   * the main thread and manages to signal this thread before the
+   * read() system call is entered, then the test condition is not
+   * exercised (no opportunity for read() to fail with EINTR). This is
+   * an acceptable race because it doesn't make the test fail, it just
+   * fails to exercise the test condition.
+   *
+   * POSIX doesn't provide us with a portable way to ensure that the
+   * thread has blocked in the system call before starting the
+   * collection; on Linux we could poll /proc/TID/stat until the
+   * thread enters state S (sleeping in an interruptible wait) but
+   * this seems like a lot of trouble for a nice-to-have feature.
+   */
+  e = read(data->fd, &c, 1);
+  asserts(e == 1, "read: %d: %s", e, strerror(errno));
+  mps_thread_dereg(thread);
+  return NULL;
+}
+
+static void test(void *stack_pointer)
+{
+  mps_arena_t arena;
+  mps_fmt_t format;
+  mps_pool_t pool;
+  int pipefd[2];
+  int e;
+  sem_t sem;
+  test_data_t data;
+  pthread_t pthread;
+
+  die(mps_arena_create_k(&arena, mps_arena_class_vm(), mps_args_none), "mps_arena_create_k");
+  die(mps_fmt_create_A(&format, arena, &fmtA), "create format");
+  MPS_ARGS_BEGIN(args) {
+     MPS_ARGS_ADD(args, MPS_KEY_FORMAT, format);
+     die(mps_pool_create_k(&pool, arena, mps_class_amc(), args), "mps_pool_create_k");
+  } MPS_ARGS_END(args);
+
+  /* Create data structures for synchronizing with the thread. */
+  e = pipe(pipefd);
+  asserts(e == 0, "pipe: %s", strerror(errno));
+  e = sem_init(&sem, 0, 0);
+  asserts(e == 0, "sem_init: %s", strerror(errno));
+  data.arena = arena;
+  data.sem = &sem;
+  data.fd = pipefd[0];
+
+  /* Start the thread. */
+  e = pthread_create(&pthread, NULL, thread_fn, &data);
+  asserts(e == 0, "pthread_create: %s", strerror(e));
+
+  /* Wait for the thread to register itself. */
+  e = sem_wait(&sem);
+  asserts(e == 0, "sem_wait: %s", strerror(errno));
+
```

Comment:
In some similar tests I have, I use a sleep of a few milliseconds here, to make it more likely that the other thread is being blocked in the `read` system call before continuing in this thread. Thus, making it "more likely" to trigger the desired problem. Sadly, it adds some time to the test cases, and does not give any strict guarantees regarding correctness.

Reply:
In practice I found that `mps_arena_collect()` has enough work to do that the `read` system call had always been entered by the time the MPS suspends threads. But it does no harm to add a millisecond of sleep here, so I did that.

Reply:
In some other work (not MPS-related) I've used the "poll /proc/PID/status until the process enters the required state" technique for synchronization in difficult cases, but that's on a project which is closely tied to Linux so that there's no portability concern.